### DISCUSSION TITLE
Substructure matching of "SMARTS on SMARTS" has unexpected results

### DISCUSSION THREAD

**Discussion by kienerj** [rdkit/rdkit#7665] [Q&A] [answered]
Simplified this would be an example what I want to do:
Check if a query molecule contains a halogen.

```
m = Chem.MolFromSmarts("[C][Cl,Br,I]")
p = Chem.MolFromSmarts("[F,Cl,Br,I]")
m.HasSubstructMatch(p)
# returns False
```

But this returns False. Ok, I can understand technically why this doesn't work hence it^s not that unexpected but more confusing.
Can this be done in any other efficient way?

Secondly I wonder if this call should potentially even raise an Exception if it doesn't really work? In more complex code it gets a bit more difficult to find that the problem is this unexpected behavior so failing with an exception would make that clear that it doesn0t work.
(Or best case: it actually works)

**Comment by greglandrum** [comment, marked as answer]
Hi @kienerj,
There are a few circumstances where query-query matching (which is REALLY hard to do correctly) recognizes valid substructure matches, but they are the exception. On the other hand, I don't think you should ever get a false positive.

The safest thing is to just assume that query-query matches will not work and avoid doing them. This is easy using the molecule's `HasQuery()` method:
```
In [8]: Chem.MolFromSmarts('C[F]').HasQuery()
Out[8]: True

In [9]: Chem.MolFromSmiles('C[F]').HasQuery()
Out[9]: False
```

**Reply by kienerj** [reply]
Yeah for this specific use-case I really need to do it, that's the issue. I have now settled on a hacky way to check the output of DescribeQuery(). it works for my limited use-case but I'm aware of this not being ideal.
